MACRO - recording driver calls as commands

MateuszMatusiak opened the pull request.

Commit:
Zapisywanie wykonanych operacji

MateuszMatusiak committed:
Merge branch 'dev-io2' into Macro

kowallus commented on src/main/java/edu/kis/powp/jobs2d/command/gui/CommandManagerWindow.java:
```diff
@@ -169,6 +169,10 @@ private void importCommandSequence() {
+	public void importSavedCommands(List<DriverCommand> commandList){
+		commandManager.setCurrentCommand(commandList, "Saved commands");
+		updatePanelPreview(commandList);
+
```
Some merge errors. Missing '}'.

kowallus commented on src/main/java/edu/kis/powp/jobs2d/features/RecordingMacro.java:
```diff
@@ -0,0 +1,52 @@
+
+	private static RecordingMacro instance;
+
+	public static RecordingMacro getInstance() {
```
add synchronized for concurrency safety

kowallus commented on src/main/java/edu/kis/powp/jobs2d/drivers/decorators/CommandUsageMonitorDecorator.java:
```diff
@@ -0,0 +1,27 @@
+
+public class CommandUsageMonitorDecorator extends BaseJob2dDriverDecorator {
+
+	RecordingMacro recordingMacro = RecordingMacro.getInstance();
```
private!

kowallus commented on src/main/java/edu/kis/powp/jobs2d/features/RecordingMacro.java:
```diff
@@ -0,0 +1,52 @@
+package edu.kis.powp.jobs2d.features;
+
+import edu.kis.powp.jobs2d.command.DriverCommand;
+import edu.kis.powp.jobs2d.command.gui.CommandManagerWindow;
```
Recording shouldn't be tangled with GUI!

kowallus replied:
I have forgotten to remove the import

kowallus commented on src/main/java/edu/kis/powp/jobs2d/features/RecordingMacro.java:
```diff
@@ -0,0 +1,52 @@
+	public void draw() {
+		CommandManagerWindow commandManager = new CommandManagerWindow(CommandsFeature.getDriverCommandManager());
+		commandManager.importSavedCommands(commandList);
+	}
```
this method doesn't and shouldn't draw. You should have method to set recorded command in DriverCommandManager instance (setCurrentCommand). You should not call GUI directly at all, and you not create new instances of already existing managers or windows.

kowallus commented on src/test/java/edu/kis/powp/jobs2d/events/SelectRecorderOptionListener.java:
```diff
@@ -0,0 +1,39 @@
+public class SelectRecorderOptionListener implements ActionListener {
+
+	RecordingMacro.RecordingOperation operation;
+	RecordingMacro recordingMacro;
```
private!

kowallus commented on src/main/java/edu/kis/powp/jobs2d/drivers/decorators/CommandUsageMonitorDecorator.java:
```diff
@@ -0,0 +1,27 @@
+import edu.kis.powp.jobs2d.command.SetPositionCommand;
+import edu.kis.powp.jobs2d.features.RecordingMacro;
+
+public class CommandUsageMonitorDecorator extends BaseJob2dDriverDecorator {
```
Why you call it "CommandUsageMonitor"? It is a recorder!

kowallus commented on src/test/java/edu/kis/powp/jobs2d/TestJobs2dApp.java:
```diff
@@ -83,13 +89,13 @@ private static void setupDrivers(Application application) {
 		driverComposite.addDriver(driver);
 
-		DriverFeature.addDriver("Special line Simulator with monitor", new Job2dDriverUsageMonitorDecorator(driver));
+		DriverFeature.addDriver("Special line Simulator with monitor", new CommandUsageMonitorDecorator(new Job2dDriverUsageMonitorDecorator(driver)));
```
Do not enforce a recording to already existing drivers. Create your own drivers.

kowallus reviewed (changes requested):
Basic idea is ok, however code doesn't run (syntax errors) and many clean code issues

Commit:
Poprawki

kowallus commented on src/test/java/edu/kis/powp/jobs2d/TestJobs2dApp.java:
```diff
@@ -89,13 +79,15 @@ private static void setupDrivers(Application application) {
 
-		DriverFeature.addDriver("Special line Simulator with monitor", new CommandUsageMonitorDecorator(new Job2dDriverUsageMonitorDecorator(driver)));
+		DriverFeature.addDriver("Special line Simulator with monitor", new Job2dDriverUsageMonitorDecorator(driver));
+		DriverFeature.addDriver("Special line Simulator with recorder", new CommandUsageRecorderDecorator(driver));
```
CommandUsageRecorderDecorator is an **empty file**, please fix it! 
I think that name CommandRecorderDriverDecorator would be better

kowallus reviewed (changes requested):
Please fix the empty file.

rafiti12 committed:
CommandUsageRecorderDecorator
- fixed the empty file
- changed CommandUsageRecorderDecorator name to CommandRecorderDriverDecorator
- made recordingMacro private

kowallus commented on src/main/java/edu/kis/powp/jobs2d/drivers/decorators/CommandRecorderDriverDecorator.java:
```diff
@@ -0,0 +1,27 @@
+
+public class CommandRecorderDriverDecorator extends BaseJob2dDriverDecorator {
+
+    private RecordingMacro recordingMacro = RecordingMacro.getInstance();
```
You should apply dependency injection to avoid fixed coupling with a single instance.

kowallus commented on src/main/java/edu/kis/powp/jobs2d/features/RecordingMacro.java:
```diff
@@ -0,0 +1,34 @@
+		return instance;
+	}
+
+	private RecordingMacro() {
+		commandList = new ArrayList<>();
+	}
```
It is not clear why RecordingMacro needs to be a singleton.

kowallus reviewed (commented):
4 pts - project final
